kod/object/active/holder/room/monsroom/elementaldungeon.kod

```diff
@@ -0,0 +1,199 @@
+% Meridian 59, Copyright 1994-2012 Andrew Kirmse and Chris Kirmse.
+% All rights reserved.
+%
+% This software is distributed under a license that is described in
+% the LICENSE file that accompanies it.
+%
+% Meridian is a registered trademark.
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+ElementalDungeon is MonsterRoom
+
+constants:
+
+   include blakston.khd
+
+resources:
+
+   room_name_elementaldungeon = "Elemental Dungeon"
+   room_elementaldungeon = elementaldungeon.roo
+   elementaldungeon_music = cave.mp3
+
+classvars:
+
+   vrName = room_name_elementaldungeon
+
+   viTeleport_row = 2
+   viTeleport_col = 2
+
+
+properties:
+
+   prRoom = room_elementaldungeon
+   piRoom_num = RID_ELEMENTAL_DUNGEON
+
+   piBaseLight = LIGHT_MIN
+   piOutside_factor = 0
+
+   prMusic = elementaldungeon_music
+
+   piGen_time = 30000
+   piGen_percent = 80
+
+   piInit_count_min = 4
+   piInit_count_max = 6
+
+   piMonster_count_max = 25
+   
+   ptDamageTimer = $
+
+messages:
+
+   Constructed()
+   {
+      propagate;
+   }
+   
+   FirstUserEntered()
+   {
+      Send(self,@StartDamageTimer);
+      Post(self,@LowerEntryway);
+      propagate;
+   }
+   
+   LastUserLeft()
+   {
+      Send(self,@CancelDamageTimer);
+      propagate;
+   }
+   
+   StartDamageTimer()
+   {
+      if ptDamageTimer = $
+      {
+         ptDamageTimer = CreateTimer(self,@DealElementalDamage,500);
+      }
+      return;
+   }
+   
+   CancelDamageTimer()
+   {
+      if ptDamageTimer <> $
+      {
+         DeleteTimer(ptDamageTimer);
+      }
+      ptDamageTimer = $;
+      return;
+   }
+   
+   DealElementalDamage(timer=$)
+   {
+      ptDamageTimer = $;
+
+      Send(self,@BurnInFire);
+      Send(self,@FreezeInWater);
+      Send(self,@ShockInAir);
+      Send(self,@ShakeInEarth);
```

Comment:
The element damaging code can likely be structured in a more efficient way, so that you only traverse the active list once, get each object's coordinates once, and depending on where they are, call the appropriate damage code. This would result in reducing the workload of this code by a factor of 4.

Also because it is a room, use First(i) instead of Send(self,@HolderExtractObject,#data=i).